pkg-config is used to find libbpf, but CFLAGS are not propagated to clang invocation

Hi,

The xdp-tools configure script has support for locating libbpf via pkg-config. Unfortunately, this support is incomplete. The CFLAGS needed for libbpf seem to be propagated correctly to the various build system targets except for the BPF byte code built with clang. This means that the required -I flags are not used, which will cause xdp-tools to fail to build if libbpf is installed outside of the system include directories that clang searches by default.

Here is an example with libbpf installed and usable via pkg-config but installed into a path under /opt:

```
root@12fcd0d03c62:/tmp/xdp-tools# pkg-config libbpf --modversion
1.3.4
root@12fcd0d03c62:/tmp/xdp-tools# export FORCE_SYSTEM_LIBBPF=1
root@12fcd0d03c62:/tmp/xdp-tools# ./configure
Found clang binary 'clang' with version 18 (from 'Ubuntu clang version 18.1.3 (1ubuntu1)')
bpftool not found or doesn't support skeleton generation; not building all tools
libbpf support: system v1.3.4
  perf_buffer__consume support: yes
  btf__load_from_kernel_by_id support: yes
  btf__type_cnt support: yes
  bpf_object__next_map support: yes
  bpf_object__next_program support: yes
  bpf_program__insn_cnt support: yes
  bpf_program__type support: yes
  bpf_program__flags support: yes
  bpf_program__expected_attach_type support: yes
  bpf_map_create support: yes
  perf_buffer__new_raw support: yes
  bpf_xdp_attach support: yes
  bpf_map__set_autocreate support: yes
  bpf_prog_test_run_opts support: yes
  bpf_xdp_query support: yes
secure_getenv support: yes
cap-ng support: no
root@12fcd0d03c62:/tmp/xdp-tools# make

lib

  libxdp
    CC       staticobjs/libxdp.o
    CC       staticobjs/xsk.o
    M4       xdp-dispatcher.c
    CLANG    xdp-dispatcher.o
xdp-dispatcher.c:4:10: fatal error: 'bpf/bpf_helpers.h' file not found
    4 | #include <bpf/bpf_helpers.h>
      |          ^~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [Makefile:138: xdp-dispatcher.o] Error 1
make[1]: *** [Makefile:20: libxdp] Error 2
make: *** [Makefile:31: lib] Error 2
root@12fcd0d03c62:/tmp/xdp-tools# make V=1

lib
make[1]: Entering directory '/tmp/xdp-tools/lib'

  libxdp
make[2]: Entering directory '/tmp/xdp-tools/lib/libxdp'
clang -target bpf -Wall -Wno-unused-value -Wno-pointer-sign -Wno-compare-distinct-pointer-types -Wno-visibility -Werror -fno-stack-protector -DBPF_DIR_MNT=\"/sys/fs/bpf\" -DBPF_OBJECT_PATH=\"/usr/local/lib/bpf\" -DMAX_DISPATCHER_ACTIONS=10 -DTOOLS_VERSION=\""1.5.3"\" -DLIBBPF_VERSION=\"1.3.4\" -DRUNDIR=\"/run\" -DHAVE_LIBBPF_PERF_BUFFER__CONSUME -DHAVE_LIBBPF_BTF__LOAD_FROM_KERNEL_BY_ID -DHAVE_LIBBPF_BTF__TYPE_CNT -DHAVE_LIBBPF_BPF_OBJECT__NEXT_MAP -DHAVE_LIBBPF_BPF_OBJECT__NEXT_PROGRAM -DHAVE_LIBBPF_BPF_PROGRAM__INSN_CNT -DHAVE_LIBBPF_BPF_PROGRAM__TYPE -DHAVE_LIBBPF_BPF_PROGRAM__FLAGS -DHAVE_LIBBPF_BPF_PROGRAM__EXPECTED_ATTACH_TYPE -DHAVE_LIBBPF_BPF_MAP_CREATE -DHAVE_LIBBPF_PERF_BUFFER__NEW_RAW -DHAVE_LIBBPF_BPF_XDP_ATTACH -DHAVE_LIBBPF_BPF_MAP__SET_AUTOCREATE -DHAVE_LIBBPF_BPF_PROG_TEST_RUN_OPTS -DHAVE_LIBBPF_BPF_XDP_QUERY -DHAVE_SECURE_GETENV -DDEBUG -DLIBBPF_DYNAMIC -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64  -I/usr/include/x86_64-linux-gnu/  -I../../headers -I/usr/include/x86_64-linux-gnu/  -O2 -c -g -o xdp-dispatcher.o xdp-dispatcher.c
xdp-dispatcher.c:4:10: fatal error: 'bpf/bpf_helpers.h' file not found
    4 | #include <bpf/bpf_helpers.h>
      |          ^~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [Makefile:138: xdp-dispatcher.o] Error 1
make[2]: Leaving directory '/tmp/xdp-tools/lib/libxdp'
make[1]: *** [Makefile:20: libxdp] Error 2
make[1]: Leaving directory '/tmp/xdp-tools/lib'
make: *** [Makefile:31: lib] Error 2
root@12fcd0d03c62:/tmp/xdp-tools#
```

I have not tracked down a proper fix but a quick-and-dirty hack to the configure script like the following worked for me to import the correct -I flags for libbpf via pkg-config:

```diff
diff --git a/configure b/configure
index 6b0fd27..354b9b9 100755
--- a/configure
+++ b/configure
@@ -128,7 +128,7 @@ check_toolchain()
     echo "PKG_CONFIG:=${PKG_CONFIG}" >>$CONFIG
     echo "CC:=${CC}" >>$CONFIG
     echo "OBJCOPY:=${OBJCOPY}" >>$CONFIG
-    echo "CLANG:=${CLANG}" >>$CONFIG
+    echo "CLANG:=${CLANG} $(${PKG_CONFIG} --cflags-only-I libbpf)" >>$CONFIG
     echo "M4:=${M4}" >>$CONFIG
     echo "EMACS:=${EMACS}" >>$CONFIG
     echo "ARCH_INCLUDES:=$ARCH_INCLUDES" >> $CONFIG
```

Here is a build of xdp-tools in an Ubuntu 24.04 container with libbpf built from git and installed into /opt/libbpf.

```
root@3c6f2e8ebf99:/tmp# git clone https://github.com/xdp-project/xdp-tools.git
Cloning into 'xdp-tools'...
remote: Enumerating objects: 7256, done.
remote: Counting objects: 100% (1504/1504), done.
remote: Compressing objects: 100% (400/400), done.
remote: Total 7256 (delta 1210), reused 1118 (delta 1100), pack-reused 5752 (from 3)
Receiving objects: 100% (7256/7256), 1.96 MiB | 20.87 MiB/s, done.
Resolving deltas: 100% (5170/5170), done.
root@3c6f2e8ebf99:/tmp# cd xdp-tools/
root@3c6f2e8ebf99:/tmp/xdp-tools# git rev-parse HEAD
9665aac5438edd4099f75b5c8275ec9a28a4f3aa
root@3c6f2e8ebf99:/tmp/xdp-tools# export PKG_CONFIG_PATH=/opt/libbpf/lib/pkgconfig
root@3c6f2e8ebf99:/tmp/xdp-tools# export FORCE_SYSTEM_LIBBPF=1
root@3c6f2e8ebf99:/tmp/xdp-tools# ./configure
Found clang binary 'clang' with version 18 (from 'Ubuntu clang version 18.1.3 (1ubuntu1)')
/usr/sbin/bpftool
bpftool not found or doesn't support skeleton generation; not building all tools
libbpf support: system v1.6.0
  perf_buffer__consume support: yes
  btf__load_from_kernel_by_id support: yes
  btf__type_cnt support: yes
  bpf_object__next_map support: yes
  bpf_object__next_program support: yes
  bpf_program__insn_cnt support: yes
  bpf_program__type support: yes
  bpf_program__flags support: yes
  bpf_program__expected_attach_type support: yes
  bpf_map_create support: yes
  perf_buffer__new_raw support: yes
  bpf_xdp_attach support: yes
  bpf_map__set_autocreate support: yes
  bpf_prog_test_run_opts support: yes
  bpf_xdp_query support: yes
secure_getenv support: yes
cap-ng support: no
root@3c6f2e8ebf99:/tmp/xdp-tools# make

lib

  libxdp
    CC       staticobjs/libxdp.o
    CC       staticobjs/xsk.o
    M4       xdp-dispatcher.c
    CLANG    xdp-dispatcher.o
xdp-dispatcher.c:4:10: fatal error: 'bpf/bpf_helpers.h' file not found
    4 | #include <bpf/bpf_helpers.h>
      |          ^~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [Makefile:138: xdp-dispatcher.o] Error 1
make[1]: *** [Makefile:20: libxdp] Error 2
make: *** [Makefile:31: lib] Error 2
root@3c6f2e8ebf99:/tmp/xdp-tools# make V=1

lib
make[1]: Entering directory '/tmp/xdp-tools/lib'

  libxdp
make[2]: Entering directory '/tmp/xdp-tools/lib/libxdp'
clang -target bpf -Wall -Wno-unused-value -Wno-pointer-sign -Wno-compare-distinct-pointer-types -Wno-visibility -Werror -fno-stack-protector -DBPF_DIR_MNT=\"/sys/fs/bpf\" -DBPF_OBJECT_PATH=\"/usr/local/lib/bpf\" -DMAX_DISPATCHER_ACTIONS=10 -DTOOLS_VERSION=\""1.5.3"\" -DLIBBPF_VERSION=\"1.6.0\" -DRUNDIR=\"/run\" -DHAVE_LIBBPF_PERF_BUFFER__CONSUME -DHAVE_LIBBPF_BTF__LOAD_FROM_KERNEL_BY_ID -DHAVE_LIBBPF_BTF__TYPE_CNT -DHAVE_LIBBPF_BPF_OBJECT__NEXT_MAP -DHAVE_LIBBPF_BPF_OBJECT__NEXT_PROGRAM -DHAVE_LIBBPF_BPF_PROGRAM__INSN_CNT -DHAVE_LIBBPF_BPF_PROGRAM__TYPE -DHAVE_LIBBPF_BPF_PROGRAM__FLAGS -DHAVE_LIBBPF_BPF_PROGRAM__EXPECTED_ATTACH_TYPE -DHAVE_LIBBPF_BPF_MAP_CREATE -DHAVE_LIBBPF_PERF_BUFFER__NEW_RAW -DHAVE_LIBBPF_BPF_XDP_ATTACH -DHAVE_LIBBPF_BPF_MAP__SET_AUTOCREATE -DHAVE_LIBBPF_BPF_PROG_TEST_RUN_OPTS -DHAVE_LIBBPF_BPF_XDP_QUERY -DHAVE_SECURE_GETENV -DDEBUG -DLIBBPF_DYNAMIC -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64  -I/usr/include/x86_64-linux-gnu/  -I../../headers -I/usr/include/x86_64-linux-gnu/  -O2 -c -g -o xdp-dispatcher.o xdp-dispatcher.c
xdp-dispatcher.c:4:10: fatal error: 'bpf/bpf_helpers.h' file not found
    4 | #include <bpf/bpf_helpers.h>
      |          ^~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [Makefile:138: xdp-dispatcher.o] Error 1
make[2]: Leaving directory '/tmp/xdp-tools/lib/libxdp'
make[1]: *** [Makefile:20: libxdp] Error 2
make[1]: Leaving directory '/tmp/xdp-tools/lib'
make: *** [Makefile:31: lib] Error 2
root@3c6f2e8ebf99:/tmp/xdp-tools# cat config.mk
# Generated config
# user can control verbosity similar to kernel builds (e.g., V=1)
ifeq ("$(origin V)", "command line")
  VERBOSE = $(V)
endif
ifndef VERBOSE
  VERBOSE = 0
endif
ifeq ($(VERBOSE),1)
  Q =
else
  Q = @
endif
ifeq ($(VERBOSE),0)
MAKEFLAGS += --no-print-directory
endif


ifeq ($(VERBOSE), 0)
    QUIET_CC       = @echo '    CC       '$@;
    QUIET_CLANG    = @echo '    CLANG    '$@;
    QUIET_LINK     = @echo '    LINK     '$@;
    QUIET_INSTALL  = @echo '    INSTALL  '$@;
    QUIET_M4       = @echo '    M4       '$@;
    QUIET_GEN      = @echo '    GEN      '$@;
endif
PRODUCTION:=0
DYNAMIC_LIBXDP:=0
MAX_DISPATCHER_ACTIONS:=10
BPF_TARGET:=bpf
PKG_CONFIG:=pkg-config
CC:=gcc
OBJCOPY:=objcopy
CLANG:=clang
M4:=m4
EMACS:=
ARCH_INCLUDES:=-I/usr/include/x86_64-linux-gnu/
READELF:=readelf
BPFTOOL:=
SYSTEM_LIBBPF:=y
LIBBPF_VERSION=1.6.0
CFLAGS +=  -I/opt/libbpf/include
LDLIBS +=  -L/opt/libbpf/lib -lbpf
OBJECT_LIBBPF =
HAVE_FEATURES+=LIBBPF_PERF_BUFFER__CONSUME
HAVE_FEATURES+=LIBBPF_BTF__LOAD_FROM_KERNEL_BY_ID
HAVE_FEATURES+=LIBBPF_BTF__TYPE_CNT
HAVE_FEATURES+=LIBBPF_BPF_OBJECT__NEXT_MAP
HAVE_FEATURES+=LIBBPF_BPF_OBJECT__NEXT_PROGRAM
HAVE_FEATURES+=LIBBPF_BPF_PROGRAM__INSN_CNT
HAVE_FEATURES+=LIBBPF_BPF_PROGRAM__TYPE
HAVE_FEATURES+=LIBBPF_BPF_PROGRAM__FLAGS
HAVE_FEATURES+=LIBBPF_BPF_PROGRAM__EXPECTED_ATTACH_TYPE
HAVE_FEATURES+=LIBBPF_BPF_MAP_CREATE
HAVE_FEATURES+=LIBBPF_PERF_BUFFER__NEW_RAW
HAVE_FEATURES+=LIBBPF_BPF_XDP_ATTACH
HAVE_FEATURES+=LIBBPF_BPF_MAP__SET_AUTOCREATE
HAVE_FEATURES+=LIBBPF_BPF_PROG_TEST_RUN_OPTS
HAVE_FEATURES+=LIBBPF_BPF_XDP_QUERY
HAVE_FEATURES += SECURE_GETENV
root@3c6f2e8ebf99:/tmp/xdp-tools#

```

I see CFLAGS being populated with the header include path of libbpf obtained from pkg-config (`CFLAGS +=  -I/opt/libbpf/include`) but I do not see `-I/opt/libbpf/include` appearing in the clang invocation:

```
clang -target bpf [...] -I/usr/include/x86_64-linux-gnu/  -I../../headers -I/usr/include/x86_64-linux-gnu/  -O2 -c -g -o xdp-dispatcher.o xdp-dispatcher.c
```

--- a/lib/defines.mk
+++ b/lib/defines.mk
@@ -43,7 +43,7 @@ endif
 DEFINES += -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64
 
 CFLAGS += -std=gnu11 -Wextra -Werror $(DEFINES) $(ARCH_INCLUDES)
-BPF_CFLAGS += $(DEFINES) $(filter -ffile-prefix-map=%,$(CFLAGS)) $(ARCH_INCLUDES)
+BPF_CFLAGS += $(DEFINES) $(filter -ffile-prefix-map=%,$(CFLAGS)) $(filter -I%,$(CFLAGS)) $(ARCH_INCLUDES)
 
 CONFIGMK := $(LIB_DIR)/../config.mk
 LIBMK := Makefile $(CONFIGMK) $(LIB_DIR)/defines.mk $(LIB_DIR)/common.mk $(LIB_DIR)/../version.mk

